lib/twilioconnection.js

```diff
@@ -0,0 +1,341 @@
+
+/**
+ * A {@link TwilioConnection} represents a WebSocket connection
+ * to a Twilio Connections Messaging Protocol (TCMP) server.
+ * @fires TwilioConnection#close
+ * @fires TwilioConnection#error
+ * @fires TwilioConnection#message
+ * @fires TwilioConnection#open
+ */
+class TwilioConnection extends StateMachine {
+  /**
+   * Construct a {@link TwilioConnection}.
+   * @param {string} serverUrl - TCMP server url
+   * @param {TwilioConnectionOptions} options - {@link TwilioConnection} options
+   */
+  constructor(serverUrl, options) {
+    super('connecting', states);
+
+    Object.defineProperties(this, {
+      _consecutiveHeartbeatsMissed: {
+        value: 0,
+        writable: true
+      },
+      _heartbeatTimeout: {
+        value: null,
+        writable: true
+      },
+      _messageQueue: {
+        value: []
+      },
+      _options: {
+        value: Object.assign({
+          maxConsecutiveMissedHeartbeats: DEFAULT_MAX_CONSECUTIVE_MISSED_HEARTBEATS,
+          requestedHeartbeatTimeout: DEFAULT_MAX_REQUESTED_HEARTBEAT_TIMEOUT,
+          welcomeTimeout: DEFAULT_WELCOME_TIMEOUT,
+          WebSocket
+        }, options)
+      },
+      _ws: {
+        value: null,
+        writable: true
+      }
+    });
+
+    this.on('stateChanged', (state, error) => ({
+      closed: () => this.emit('close', error),
+      open: () => this.emit('open')
+    }[state]()));
+
+    this._connect(serverUrl);
+  }
+
+  /**
+   * The number of consecutive "hearbeat" messages missed.
+   * @property {number}
+   */
+  get consecutiveHeartbeatsMissed() {
+    return this._consecutiveHeartbeatsMissed;
+  }
+
+  /**
+   * Connect to the TCMP server.
+   * @param {string} serverUrl
+   * @private
+   */
+  _connect(serverUrl) {
+    this._ws = new this._options.WebSocket(serverUrl);
+    const ws = this._ws;
+
+    ws.addEventListener('close', event => {
+      this._messageQueue.splice(0);
+      this.transition('closed', null, event.code !== WS_CLOSE_NORMAL
+        ? new Error(`WebSocket Error ${event.code}: ${event.reason}`)
+        : null);
+    });
+
+    ws.addEventListener('message', message => {
+      try {
+        message = JSON.parse(message.data);
+      } catch (error) {
+        this.emit('error', error);
+        return;
+      }
+      switch (message.type) {
+        case 'bad':
+          this._handleBad(message);
+          break;
+        case 'heartbeat':
+          this._handleHeartbeat();
+          break;
+        case 'msg':
+          this._handleMessage(message);
+          break;
+        case 'welcome':
+          this._handleWelcome(message);
+          break;
+        default:
+          this.emit('error', new Error(`Unknown message type: ${message.type}`));
+          break;
+      }
+    });
+
+    ws.addEventListener('open', () => {
+      this._sendHello();
+      const { welcomeTimeout } = this._options;
+      setTimeout(() => this._handleWelcomeTimeout(), welcomeTimeout);
```

Comment:
I think we should clean up this timeout if the TwilioConnection is closed or a welcome is received, as open handles like the timeouts created by `setTimeout` will keep the Node process open